### PR TITLE
Modify jwt authentication, adapt to BC

### PR DESCRIPTION
**Pull request by lixiumei123**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -200,8 +200,8 @@ file(COPY ${CMAKE_SOURCE_DIR}/neuron.yaml DESTINATION ${CMAKE_BINARY_DIR}/config
 file(COPY ${CMAKE_SOURCE_DIR}/plugins/ekuiper/ekuiper.json DESTINATION ${CMAKE_BINARY_DIR}/plugins/schema/)
 file(COPY ${CMAKE_SOURCE_DIR}/plugins/modbus/modbus-tcp.json DESTINATION ${CMAKE_BINARY_DIR}/plugins/schema/)
 file(COPY ${CMAKE_SOURCE_DIR}/plugins/mqtt/mqtt.json DESTINATION ${CMAKE_BINARY_DIR}/plugins/schema/)
-file(COPY ${CMAKE_SOURCE_DIR}/private.key DESTINATION ${CMAKE_BINARY_DIR}/config)
-file(COPY ${CMAKE_SOURCE_DIR}/public.pem DESTINATION ${CMAKE_BINARY_DIR}/config)
+file(COPY ${CMAKE_SOURCE_DIR}/Neuron.key DESTINATION ${CMAKE_BINARY_DIR}/config)
+file(COPY ${CMAKE_SOURCE_DIR}/Neuron.pem DESTINATION ${CMAKE_BINARY_DIR}/config)
```
Comment (fengzeroz) [inline CMakeLists.txt:204]: rename neuron.key neuron.pem

File: package.sh [hunk 2]
```diff
@@ -48,8 +48,8 @@ cp build/libneuron-base.so \
     ${package_name}
 
 cp build/config/neuron.yaml \
-    private.key \
-    public.pem \
+    Neuron.key \
+    Neuron.pem \
```
Comment (fengzeroz) [inline package.sh:52]: rename neuron

File: ft/normal_test.robot [hunk 3]
```diff
@@ -54,10 +54,10 @@ logout with valid token, it should return success
 
 *** Keywords ***
 Neuron Context Ready
-  Import Neuron API Resource
-  Skip If Not Http API
+    Import Neuron API Resource
+    Skip If Not Http API
 
-  Neuron Ready
+    Neuron Ready
```
Comment (fengzeroz) [inline ft/normal_test.robot:60]: format

File: package/create_neuron_deb.py [hunk 4]
```diff
@@ -18,8 +18,8 @@
 rules.append(mkdeb.FileMap("../.gitkeep", "/opt/neuron/logs/"))
 rules.append(mkdeb.FileMap("../.gitkeep", "/opt/neuron/core/"))
 
-rules.append(mkdeb.FileMap("../private.key", "/opt/neuron/config/"))
-rules.append(mkdeb.FileMap("../public.pem", "/opt/neuron/config/"))
+rules.append(mkdeb.FileMap("../Neuron.key", "/opt/neuron/config/"))
+rules.append(mkdeb.FileMap("../Neuron.pem", "/opt/neuron/config/"))
```
Comment (fengzeroz) [inline package/create_neuron_deb.py:22]: neuron

File: tests/CMakeLists.txt [hunk 5]
```diff
@@ -140,8 +140,8 @@ target_include_directories(persistence_test PRIVATE
 )
 target_link_libraries(persistence_test neuron-base gtest_main gtest)
 
-file(COPY ${CMAKE_SOURCE_DIR}/private.key DESTINATION ${CMAKE_BINARY_DIR}/tests)
-file(COPY ${CMAKE_SOURCE_DIR}/public.pem DESTINATION ${CMAKE_BINARY_DIR}/tests)
+file(COPY ${CMAKE_SOURCE_DIR}/Neuron.key DESTINATION ${CMAKE_BINARY_DIR}/tests/config)
+file(COPY ${CMAKE_SOURCE_DIR}/Neuron.pem DESTINATION ${CMAKE_BINARY_DIR}/tests/config)
```
Comment (fengzeroz) [inline tests/CMakeLists.txt:144]: neuron

File: src/utils/neu_jwt.h [hunk 6]
```diff
@@ -25,7 +25,13 @@ extern "C" {
 
 #include <jwt.h>
 
-int  neu_jwt_init(char *pub_key_path, char *priv_key_path);
+typedef struct {
+    char   name[256];
+    char   value[4096];
+    size_t len;
+} key_file_t;
```
Comment (fengzeroz) [inline src/utils/neu_jwt.h:32]: put key_file_t to neu_jwt.c

File: src/utils/neu_jwt.c [hunk 7]
```diff
@@ -27,45 +29,119 @@
 #include "log.h"
 #include "neu_jwt.h"
 
-char   public_key[4096]  = { 0 };
-char   private_key[4096] = { 0 };
-char   token_local[4096] = { 0 };
+char token_local[4096] = { 0 };
+
+int        n_pub_file = 0;
+key_file_t pub_key_files[256];
```
Comment (fengzeroz) [inline src/utils/neu_jwt.c:35]: max size 16

File: src/utils/neu_jwt.c [hunk 8]
```diff
@@ -27,45 +29,119 @@
 #include "log.h"
 #include "neu_jwt.h"
 
-char   public_key[4096]  = { 0 };
-char   private_key[4096] = { 0 };
-char   token_local[4096] = { 0 };
+char token_local[4096] = { 0 };
```
Comment (fengzeroz) [inline src/utils/neu_jwt.c:32]: del

File: src/utils/neu_jwt.c (hunk 7 above)
Comment (fengzeroz) [inline src/utils/neu_jwt.c:35]: put pub_key_files in plugin struct

File: src/utils/neu_jwt.c [hunk 9]
```diff
@@ -27,45 +29,119 @@
...
+    fp_pub_key = fopen(pub_path, "r");
+
+    if (fp_pub_key == NULL) {
+        log_error("Failed to open public key file: %s, errno: %d", public_name,
+                  errno);
+        return -1;
+    }
+
+    pub_key_files[n_pub_file].len =
+        fread(pub_key_string, 1, sizeof(pub_key_string), fp_pub_key);
```
Comment (fengzeroz) [inline src/utils/neu_jwt.c:59]: fread(pub_key_files[n_pub_file].value

File: src/utils/neu_jwt.c [hunk 10]
```diff
@@ -27,45 +29,119 @@
...
                   errno);
         return -1;
     }
+
+    strncpy(pub_key_files[n_pub_file].value, pub_key_string,
+            sizeof(pub_key_files[n_pub_file].value));
+
+    memcpy(pub_key_files[n_pub_file].name, public_name,
+           strlen(public_name) - 4);
+
```
Comment (fengzeroz) [inline src/utils/neu_jwt.c:71]: strncpy

Comment (fengzeroz) [reply src/utils/neu_jwt.c:71]: why -4

File: src/utils/neu_jwt.c [hunk 11]
```diff
@@ -84,11 +160,29 @@ int neu_jwt_new(char **token)
 
     gettimeofday(&tv, NULL);
 
-    jwt_add_grant_int(jwt, "exp", tv.tv_sec + 60 * 60);
+    ret = jwt_add_grant(jwt, "iss", "Neuron");
```
Comment (fengzeroz) [inline src/utils/neu_jwt.c:163]: neuron

File: src/utils/neu_jwt.c [hunk 12]
```diff
@@ -84,11 +160,29 @@ int neu_jwt_new(char **token)
...
+    if (ret != 0) {
+        goto err_out;
+    }
+
+    ret = jwt_add_grant_int(jwt, "exp", tv.tv_sec + 60 * 60);
+    if (ret != 0) {
+        goto err_out;
+    }
+
+    ret = jwt_add_grant(jwt, "aud", "Neuron");
```
Comment (fengzeroz) [inline src/utils/neu_jwt.c:178]: neuron

File: src/utils/neu_jwt.c [hunk 13]
```diff
@@ -188,13 +313,15 @@ int neu_jwt_validate(char *b_token)
...
     }
 
-    jwt_valid_free(jwt_valid);
-    jwt_free(jwt);
-
     if (strcmp(token_local, token) != 0) {
+        jwt_valid_free(jwt_valid);
+        jwt_free(jwt);
         return NEU_ERR_INVALID_TOKEN;
     }
```
Comment (fengzeroz) [inline src/utils/neu_jwt.c:320]: remove

File: src/utils/neu_jwt.c (hunk 10 above)
Comment (lixiumei123) [reply src/utils/neu_jwt.c:71]: public_name is the name of the public key file, I need the name before the ".pem"

File: src/utils/neu_jwt.c [hunk 14]
```diff
@@ -27,45 +29,119 @@
...
+        log_error("Failed to open private key file: %s, errno: %d",
+                  private_name, errno);
+        return -1;
+    }
+
     priv_key_len = fread(private_key, 1, sizeof(private_key), fp_priv_key);
     if (priv_key_len <= 0) {
-        log_error("Failed to read key file: %s, error: %d", priv_key_path,
-                  errno);
+        log_error("Failed to read key file: %s, error: %d", priv_path, errno);
```
Comment (eeff) [inline src/utils/neu_jwt.c:95]: bug: need to close `fp_priv_key` before return.

File: src/utils/neu_jwt.c [hunk 15]
```diff
@@ -27,45 +29,119 @@
...
         return -1;
     }
+
+    pub_key_files[n_pub_file].len =
+        fread(pub_key_files[n_pub_file].key, 1,
+              sizeof(pub_key_files[n_pub_file].key), fp_pub_key);
+    if (pub_key_files[n_pub_file].len <= 0) {
+        log_error("Failed to read public key file: %s, errno: %d", public_name,
+                  errno);
+        return -1;
```
Comment (eeff) [inline src/utils/neu_jwt.c:66]: bug: need to close `fp_pub_key` before return.

**Pull request merged**